libc-top-half/musl/src/thread/wasm32/wasi_thread_start.s

```diff
@@ -28,4 +28,20 @@ wasi_thread_start:
 	local.get   1  # start_arg
 	call __wasi_thread_start_C
 
+	# Unlock thread list. (as CLONE_CHILD_CLEARTID would do for Linux)
+	#
+	# Note: once we unlock the thread list, our "map_base" can be freed
+	# by a joining thread. It's safe as we are in ASM and no longer use
+	# our C stack or pthread_t.
+	i32.const   __thread_list_lock
+	i32.const   0
+	i32.atomic.store 0
+	# As an optimization, we can check tl_lock_waiters here.
+	# But for now, simply wake up unconditionally as
+	# CLONE_CHILD_CLEARTID does.
+	i32.const   __thread_list_lock
+	i32.const   1
+	memory.atomic.notify 0
+	drop
+
```

Comment:
Reading over the definition of `__tl_unlock` it additionally decrements `tl_lock_count`. It might be good to assert that is zero in C before returning here to perform this. Additionally it might be good to leave a comment in `__tl_unlock` that this needs updating if the implementation changes.

Reply:
Why not just `call __tl_unlock`?

Reply:
> Reading over the definition of `__tl_unlock` it additionally decrements `tl_lock_count`. It might be good to assert that is zero in C before returning here to perform this. Additionally it might be good to leave a comment in `__tl_unlock` that this needs updating if the implementation changes.

i feel it belongs to upstream musl, not here.

Reply:
> Why not just `call __tl_unlock`?

because `__tl_unlock` is a C function, which potentially requires C stack.

Reply:
I'm not sure what you mean by it belongs in upstream musl? Wasi-libc is already patching musl a fair bit, so what I'm saying is that as part of the patching the cases where `__pthread_exit` returns should all assert that the `tl_lock_count` is zero.

Reply:
> I'm not sure what you mean by it belongs in upstream musl? Wasi-libc is already patching musl a fair bit, so what I'm saying is that as part of the patching the cases where `__pthread_exit` returns should all assert that the `tl_lock_count` is zero.

i meant tl_lock_count should be zero when a thread exits is not wasi-specific.
if you feel strongly i can add such an assertion.